{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a77371e3-3341-49c2-9bf3-d509a3098357",
   "metadata": {},
   "source": [
    "# Escape at key sites: logo plots and binding / escape correlations\n",
    "Make logo plots of serum escape at key sites, and look at relationship between escape and other phenotypes like ACE2 binding.\n",
    "\n",
    "First get input files / parameters from `papermill` and import Python modules:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a02e531e-d12b-4211-8fc0-f75f723d8f6f",
   "metadata": {
    "editable": true,
    "slideshow": {
     "slide_type": ""
    },
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "# this cell is tagged as `parameters` for papermill parameterization\n",
    "dms_csv = None\n",
    "per_antibody_csv = None\n",
    "pango_consensus_seqs_json = None\n",
    "codon_seq = None\n",
    "logoplot_subdir = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0c57912e-00ce-4af5-a86d-8338131c1870",
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools\n",
    "import os\n",
    "\n",
    "import altair as alt\n",
    "\n",
    "import Bio.SeqIO\n",
    "\n",
    "import dmslogo\n",
    "import dmslogo.colorschemes\n",
    "\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import numpy\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "_ = alt.data_transformers.disable_max_rows()\n",
    "\n",
    "os.makedirs(logoplot_subdir, exist_ok=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a78bee81-95fa-4cc0-9551-61279fd5fa55",
   "metadata": {},
   "source": [
    "## Read input data\n",
    "Keep only mutations with all phenotypes measured:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5d29324-b3ec-42f3-ae23-94ee154a0935",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read averages for all DMS measurements\n",
    "dms_df = (\n",
    "    pd.read_csv(dms_csv)\n",
    "    .rename(\n",
    "        columns={\"human sera escape\": \"sera escape\", \"spike mediated entry\": \"cell entry\"}\n",
    "    )\n",
    "    .query(\"`sera escape`.notnull() and `cell entry`.notnull() and `ACE2 binding`.notnull()\")\n",
    ")\n",
    "\n",
    "# read per antibody values, merge with averages to create escape_df\n",
    "per_antibody_df = pd.read_csv(per_antibody_csv)\n",
    "\n",
    "assert per_antibody_df[\"antibody_set\"].nunique() == 1, \"code expects 1 antibody_set\"\n",
    "\n",
    "if (\n",
    "    (intersection := set(dms_df.columns).intersection(per_antibody_df.columns))\n",
    "    != {\"site\", \"wildtype\", \"mutant\"}\n",
    "):\n",
    "    raise ValueError(f\"unexpected {intersection=}\")\n",
    "\n",
    "assert \"average\" not in per_antibody_df[\"antibody\"]\n",
    "\n",
    "escape_df = (\n",
    "    pd.concat(\n",
    "        [\n",
    "            dms_df[[\"site\", \"wildtype\", \"mutant\", \"sera escape\"]].rename(\n",
    "                columns={\"sera escape\": \"escape\"}\n",
    "            ).assign(antibody=\"average\"),\n",
    "            per_antibody_df.drop(columns=\"antibody_set\"),\n",
    "        ],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "    .merge(dms_df.drop(columns=\"sera escape\"), validate=\"many_to_one\")\n",
    "    .assign(wildtype_site=lambda x: x[\"wildtype\"] + x[\"site\"].astype(str))\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b13a46e-10a4-4e64-8c5c-93e8c72b1677",
   "metadata": {},
   "source": [
    "## Determine key sites of strongest escape\n",
    "Get key sites with most site escape, and plot their site escape values in interactive chart:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9297536-f262-45b5-8cab-631699275ed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specification of how to choose sites\n",
    "key_sites_by_rank = {\n",
    "    \"total_positive_escape\": {\n",
    "        \"any antibody\": 5,\n",
    "        \"average of antibodies\": 15,\n",
    "    },\n",
    "}\n",
    "# sites used in neuts\n",
    "key_sites_manual = []\n",
    "\n",
    "# get total magnitude of escape at each site, both for averages\n",
    "# and across all individual antibodies\n",
    "site_escape_df = (\n",
    "    escape_df\n",
    "    .assign(\n",
    "        is_average=lambda x: numpy.where(\n",
    "            x[\"antibody\"] == \"average\", \"average of antibodies\", \"any antibody\"),\n",
    "    )\n",
    "    .groupby([\"is_average\", \"antibody\", \"site\", \"sequential_site\"], as_index=False)\n",
    "    .aggregate(\n",
    "        total_abs_escape=pd.NamedAgg(\"escape\", lambda s: s.abs().sum()),\n",
    "        total_positive_escape=pd.NamedAgg(\"escape\", lambda s: s.clip(lower=0).sum()),\n",
    "        total_negative_escape=pd.NamedAgg(\"escape\", lambda s: s.clip(upper=0).abs().sum()),\n",
    "    )\n",
    "    .groupby([\"is_average\", \"site\", \"sequential_site\"], as_index=False)\n",
    "    .aggregate(\n",
    "        {\n",
    "            \"total_abs_escape\": \"max\",\n",
    "            \"total_positive_escape\": \"max\",\n",
    "            \"total_negative_escape\": \"max\",\n",
    "        }\n",
    "    )\n",
    "    .melt(\n",
    "        id_vars=[\"is_average\", \"site\", \"sequential_site\"],\n",
    "        var_name=\"site metric\",\n",
    "        value_name=\"site escape\",\n",
    "    )\n",
    "    .assign(\n",
    "        rank=lambda x: (\n",
    "            x.groupby([\"is_average\", \"site metric\"])\n",
    "            [\"site escape\"]\n",
    "            .rank(ascending=False, method=\"min\")\n",
    "            .astype(int)\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "# get key sites\n",
    "print(f\"Keeping the following manually specified sites: {key_sites_manual}\")\n",
    "key_sites = set(key_sites_manual)\n",
    "for site_metric, site_metric_d in key_sites_by_rank.items():\n",
    "    for is_average, rank in site_metric_d.items():\n",
    "        new_sites = set(\n",
    "            site_escape_df\n",
    "            .query(\"`site metric` == @site_metric\")\n",
    "            .query(\"is_average == @is_average\")\n",
    "            .query(\"rank <= @rank\")\n",
    "            [\"site\"]\n",
    "        )\n",
    "        print(f\"Adding sites with {site_metric} / {is_average} rank <= {rank}: {new_sites}\")\n",
    "        key_sites = key_sites.union(new_sites)\n",
    "print(f\"Overall keeping the following {len(key_sites)} sites: {key_sites}\")\n",
    "\n",
    "site_escape_df[\"key_site\"] = site_escape_df[\"site\"].isin(key_sites)\n",
    "\n",
    "# plot sites being kept\n",
    "site_selection = alt.selection_point(fields=[\"site\"], on=\"mouseover\", empty=False)\n",
    "\n",
    "site_metric_selection = alt.selection_point(\n",
    "    fields=[\"site metric\"],\n",
    "    value=\"total_positive_escape\",\n",
    "    bind=alt.binding_select(\n",
    "        name=\"site metric\",\n",
    "        options=site_escape_df[\"site metric\"].unique(),\n",
    "    ),\n",
    ")\n",
    "\n",
    "site_escape_chart = (\n",
    "    alt.Chart(site_escape_df)\n",
    "    .add_params(site_selection, site_metric_selection)\n",
    "    .transform_filter(site_metric_selection)\n",
    "    .encode(\n",
    "        alt.X(\"site\", sort=alt.SortField(\"sequential_site\"), scale=alt.Scale(nice=False, zero=False)),\n",
    "        alt.Y(\"site escape\"),\n",
    "        alt.Color(\"key_site\"),\n",
    "        alt.Row(\"is_average\", title=None),\n",
    "        tooltip=[alt.Tooltip(c, format=\".2f\") if site_escape_df[c].dtype == float else c for c in site_escape_df.columns],\n",
    "        strokeWidth=alt.condition(site_selection, alt.value(2), alt.value(0)),\n",
    "        opacity=alt.condition(site_selection, alt.value(1), alt.value(0.35)),\n",
    "        size=alt.condition(site_selection, alt.value(70), alt.value(30)),\n",
    "    )\n",
    "    .mark_circle(stroke=\"black\")\n",
    "    .configure_axis(grid=False)\n",
    "    .resolve_scale(y=\"independent\")\n",
    "    .properties(\n",
    "        width=600,\n",
    "        height=150,\n",
    "        title=\"Escape at each site for average of antibodies or max for any antibody\",\n",
    "    )\n",
    ")\n",
    "\n",
    "site_escape_chart"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3e4eacc-2fa5-4675-9569-8ad020198b8b",
   "metadata": {},
   "source": [
    "## Draw logo plots for key sites of strongest escape\n",
    "\n",
    "First get key sites and assign colors by ACE2 affinity:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a2cd3c7-20e2-43af-9732-c6e53a803664",
   "metadata": {},
   "outputs": [],
   "source": [
    "key_sites_df = (\n",
    "    escape_df\n",
    "    .query(\"site in @key_sites\")\n",
    "    .query(\"mutant not in ['*', '-']\")\n",
    ")\n",
    "\n",
    "# for coloring by ACE2\n",
    "ace2_colormap = dmslogo.colorschemes.ValueToColorMap(\n",
    "    minvalue=max(-2, key_sites_df[\"ACE2 binding\"].min()),\n",
    "    maxvalue=0,\n",
    "    cmap=\"YlOrBr\",\n",
    ")\n",
    "\n",
    "key_sites_df[\"color\"] = (\n",
    "    key_sites_df[\"ACE2 binding\"]\n",
    "    .clip(lower=ace2_colormap.minvalue, upper=ace2_colormap.maxvalue)\n",
    "    .map(ace2_colormap.val_to_color)\n",
    ")\n",
    "\n",
    "for orientation in [\"horizontal\", \"vertical\"]:\n",
    "    fig, _ = ace2_colormap.scale_bar(\n",
    "        orientation=orientation, label=\"ACE2 binding\",\n",
    "    )\n",
    "    display(fig)\n",
    "    svg = os.path.join(logoplot_subdir, f\"key_sites_ace2_scalebar_{orientation}.svg\")\n",
    "    print(f\"Saving to {svg}\")\n",
    "    fig.savefig(svg, bbox_inches=\"tight\")\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cbe26f2-f257-41ca-949e-780fb499cbf0",
   "metadata": {},
   "source": [
    "Get which of the key sites are single nucleotide accessible:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28f73a61-5afe-4616-9379-8523121a73ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "codon_to_aas = {}\n",
    "nts = \"ACGT\"\n",
    "for nt1, nt2, nt3 in itertools.product(nts, nts, nts):\n",
    "    codon = f\"{nt1}{nt2}{nt3}\"\n",
    "    codon_to_aas[codon] = set()\n",
    "    for i in range(len(codon)):\n",
    "        for nt in nts:\n",
    "            mutcodon = codon[: i] + nt + codon[i + 1: ]\n",
    "            aa = str(Bio.Seq.Seq(mutcodon).translate())\n",
    "            codon_to_aas[codon].add(aa)\n",
    "\n",
    "gene = str(Bio.SeqIO.read(codon_seq, \"fasta\").seq).upper()\n",
    "\n",
    "key_sites_df = (\n",
    "    key_sites_df\n",
    "    .assign(\n",
    "        codon=lambda x: x[\"sequential_site\"].map(\n",
    "            lambda r: gene[3 * (r - 1): 3 * r]\n",
    "        ),\n",
    "        codon_translated=lambda x: x[\"codon\"].map(\n",
    "            lambda c: str(Bio.Seq.Seq(c).translate())\n",
    "        ),\n",
    "        single_nt_accessible=lambda x: x.apply(\n",
    "            lambda r: r[\"mutant\"] in codon_to_aas[r[\"codon\"]],\n",
    "            axis=1,\n",
    "        )\n",
    "    )\n",
    ")\n",
    "\n",
    "key_sites_df = pd.concat(\n",
    "    [\n",
    "        key_sites_df.query(\"single_nt_accessible\").assign(\n",
    "            single_nt_accessible=\"single-nucleotide accessible\"\n",
    "        ),\n",
    "        key_sites_df.assign(single_nt_accessible=\"all measured mutations\"),\n",
    "    ],\n",
    "    ignore_index=True,\n",
    ")\n",
    "\n",
    "assert (key_sites_df[\"wildtype\"] == key_sites_df[\"codon_translated\"]).all()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a09bf3c7-c2e0-4ddf-9e1f-50940675aac1",
   "metadata": {},
   "source": [
    "Plots for averages across sera, for all mutations and just single-nucleotide accessible ones:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4538881-e728-468a-8940-31a5fcbe95a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "draw_logo_kwargs={\n",
    "    \"letter_col\": \"mutant\",\n",
    "    \"color_col\": \"color\",\n",
    "    \"xtick_col\": \"wildtype_site\",\n",
    "    \"letter_height_col\": \"escape\",\n",
    "    \"xlabel\": \"\",\n",
    "    \"clip_negative_heights\": True,\n",
    "}\n",
    "\n",
    "fig, _ = dmslogo.facet_plot(\n",
    "    data=key_sites_df.query(\"antibody == 'average'\"),\n",
    "    x_col=\"sequential_site\",\n",
    "    show_col=None,\n",
    "    gridrow_col=\"single_nt_accessible\",\n",
    "    share_ylim_across_rows=False,\n",
    "    hspace=0.6,\n",
    "    height_per_ax=2.2,\n",
    "    draw_logo_kwargs=draw_logo_kwargs,\n",
    ")\n",
    "\n",
    "display(fig)\n",
    "svg = os.path.join(logoplot_subdir, \"avg_sera_escape_at_key_sites.svg\")\n",
    "print(f\"Saving to {svg}\")\n",
    "fig.savefig(svg, bbox_inches=\"tight\")\n",
    "plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a8150839-b653-417b-b12d-7505f4f5ff14",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-10-20T21:42:14.392022Z",
     "iopub.status.busy": "2023-10-20T21:42:14.391459Z",
     "iopub.status.idle": "2023-10-20T21:42:14.400118Z",
     "shell.execute_reply": "2023-10-20T21:42:14.398892Z",
     "shell.execute_reply.started": "2023-10-20T21:42:14.391982Z"
    }
   },
   "source": [
    "Now make plots for all individual sera, both with all and only single-nucleotide accessible mutations:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3cb23a11-3b95-4634-974a-53f8c7cdd63d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for single_nt_accessible, df in key_sites_df.groupby(\"single_nt_accessible\"):\n",
    "    print(f\"\\n{single_nt_accessible=}\")\n",
    "    fig, axes = dmslogo.facet_plot(\n",
    "        data=df.query(\"antibody != 'average'\"),\n",
    "        x_col=\"sequential_site\",\n",
    "        show_col=None,\n",
    "        gridrow_col=\"antibody\",\n",
    "        share_ylim_across_rows=False,\n",
    "        hspace=0.6,\n",
    "        height_per_ax=2,\n",
    "        draw_logo_kwargs=draw_logo_kwargs,\n",
    "    )\n",
    "    display(fig)\n",
    "    svg = os.path.join(\n",
    "        logoplot_subdir,\n",
    "        f\"all_sera_escape_at_key_sites_{single_nt_accessible.replace(' ', '_')}.svg\",\n",
    "    )\n",
    "    print(f\"Saving to {svg}\")\n",
    "    fig.savefig(svg, bbox_inches=\"tight\")\n",
    "    plt.close(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc441f61-bce0-4017-baa2-3ebc15127c3b",
   "metadata": {},
   "source": [
    "## Sites of neutralization assay mutations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "966b08ec-5e08-4069-ace8-933f885b60c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "sites_in_neuts = [\n",
    "    42,\n",
    "    200,\n",
    "    234,\n",
    "    357,\n",
    "    403,\n",
    "    405,\n",
    "    420,\n",
    "    444,\n",
    "    455,\n",
    "    456,\n",
    "    473,\n",
    "    572,\n",
    "    852,\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e8047330-a400-4021-9f77-da15dc8df7a0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
